File: 프로그래머스/lv3/72413. 합승 택시 요금/합승 택시 요금.py
def solution(n, s, a, b, fares):
    edges = [[float('inf')] * n for _ in range(n)]
    for n1, n2, cost in fares:
        edges[n1 - 1][n2 - 1] = cost
        edges[n2 - 1][n1 - 1] = cost
    
    edges = floydWarshall(edges)
    cost = edges[s - 1][a - 1] + edges[s - 1][b - 1]
    for node in range(n):
        if node != s - 1:
            cost = min(cost, edges[s - 1][node] + edges[node][a - 1] + edges[node][b - 1])
    return cost


def floydWarshall(base):
    d = base[:]
    for k in range(len(base)):
        for i in range(len(base)):
            for j in range(len(base)):
                if d[i][j] > d[i][k] + d[k][j]:
                    d[i][j] = d[i][k] + d[k][j]
    for i in range(len(base)):
        d[i][i] = 0
    return d

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.kt
class Solution {
    
    val INF = 1e8.toInt()
    val edges = Array(201) { IntArray(201) { INF } }
    
    fun solution(n: Int, s: Int, a: Int, b: Int, fares: Array<IntArray>): Int {
        for (u in 1..n) {
            edges[u][u] = 0
        }
        
        for ((u, v, w) in fares) {
            edges[u][v] = w
            edges[v][u] = w
        }
        
        for (k in 1..n) {
            for (u in 1..n) {
                for (v in 1..n) {
                    edges[u][v] = minOf(edges[u][v], edges[u][k] + edges[k][v])
                }
            }
        }
        
        var answer = INF
        
        for (k in 1..n) {
            answer = minOf(answer, edges[s][k] + edges[k][a] + edges[k][b])
        }
        
        return answer
    }
}
